docs/lesson-plans/week-5.md

```diff
@@ -0,0 +1,639 @@
+    public PostsRouter(IPostsProcessor processor, Externals override) {
+        this.processor = processor;
+        this.externs = override;
+    }
+
+    public static class Externals {
+        public Router getRouter(Vertx vertx) {
+            return Router.router(vertx);
+        }
+    }
+
+    @Override
+    public Router initializeRouter(Vertx vertx) {
+        Router router = externs.getRouter(vertx);
+    ...
+```
+
+And this is how we override that behavior in the test class:
+```java
+public class PostsRouterTest {
+    ...
+
+    // Create the new externals, which overides the old behavior.
+    public static class TestExterns extends Externals {
+        public Router getRouter(Vertx vertx) {
+            return myTestRouter;
+        }
+    }
+
+    @BeforeEach
+    public void setup() {
+        ...
+        this.processor = new Processor(this.db, new TestExterns());
+        ...
+    }
+}
+```
+
+#### Exceptions
+
+When testing exceptions, JUnit provides methods and annotations to help, but we find it more helpful to use the 
+try/catch way of doing things. We think it allows you to test things about the exception better.
+
+Example:
+```java 
+try {
+    thingThatThrowsException();
+    fail(); // fail this test if we make it this far
+}
+catch (TheExpectedExceptionType e) {
+    assertEquals("My exception message", e.getMessage());
+    assertEquals(otherThingsAboutExceptionIfYouWant, e.doSomething());
+}
+```
+
+### Testing Using Mockito
+
+Mockito is a really useful library for mocking the interactions between objects that you don't have control over
+(like dependencies). It lets you create fake versions of objects which you have complete control over, and it even lets
+you verify things about how it was called.
+
+If you're interested in learning more about Mockito, [here is the 
+documentation](http://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html). It even lets you do some 
+really cool things like `spy()`, where you can pass in an existing object, have it do _most_ of the things it usually 
+does, and then just override/verify/mock what you want!
+
+#### Mocking an Object and Overriding Returns
+
+To mock an object, all you have to do is call (the static method) `mock(Thing.class)`, which will return a fake
+instance of that `Thing` class/interface. By default, it will return null 
+for every non-void method you call, but you can override that using `thenReturn`, `thenThrow`, `then` (`thenAnswer`, 
+allows you to do custom stuff), or `doReturn`, `doThrow`, `doAnswer`, `doNothing`, `doCallRealMethod` in the case of 
+void returns. You can also chain them together to have it return/throw/do things one after the other in the order 
+they're called.
+
+>An 'instance' of an object is a constructed value of the object created by its constructor. So for the class `Thing`, 
+>it's the difference between referring to `Thing.staticMethod()` as a static method and `new Thing().instanceMethod()`.
+
```

Comment:
Add a side note explaining why you might want to you thenReturn vs doReturn. I.e. use doReturn to avoid side effects
http://sangsoonam.github.io/2019/02/04/mockito-doreturn-vs-thenreturn.html#:~:text=In%20Mockito%2C%20a%20side%20effect%20happens%20to%20the%20real%20method%20invocation.&text=When%20you%20call%20a%20method,thenReturn%20has%20a%20type%20safety.

Reply:
I threw the link in the doc, but I think I'm still going to recommend thenReturn since it provides typing